_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+familiar elements from regular Java already, like package name and import statements.
+Further decomposing it reveals lack of semicolons in the end of the lines and differences
+in variable and function declarations. But also some obvious similarities like code block
+separation by usage of curly brackets and support for annotations.
+
+__Variable and function declaration__
+
+In our sample code we can see that declaring a value can be as simple as
+
+```groovy
+val reference = "my string"
+```
+
+where the type of the object that is assigned is implicitly guessed by compiler.
+Or we can explicitly declare the type of the object by writing
+
+```groovy
+val formatter: Formatter = Formatter()
+```
+
+where the type information follows variable name with colon and precedes the value
+to be assigned.
+
+Kotlin also provides an easy way to distinguish read-only and mutable fields.
+It does so by exposing two types of keywords: `val` for declaring read-only fields
+and `var` for variables that ought to be mutable. In the realm of Java, `val` corresponds
+to the `final` modifier.
+
+> *It is also worth noting how we can assign Java object to a Kotlin variable without
+any further ado.*
+
+Functions are defined using `fun` keyword followed by the function name and parameters
+in parentheses. Return type and argument types are specified using the colon notation
+just as it is with variable declaration. But as Kotlin has first class support for
+user defined **infix** functions, as seen above, we can enrich the standard library
+with syntactic sugar of our choosing with no effort whatsoever. For example here we
+have replaced
+
+```java
+assertEquals(expected, actual)
+```
+
+checks with
+
+```java
+actual equals expected
+```
+
+statements for readability, utilizing our custom infix procedure `equals`.
+
+```java
+infix fun Any.equals(expected: Any) {
+    assertEquals(expected, this)
+}
+```
+
+Of course, all other gems of Kotlin are readily available to us, including
+convenient array declaration, avoidance of clumsy `for` loops using `map` and
+other functional elements.
+
+Last but not least, we annotate out test functions with `@Test` as usual to make
+compiler aware of them.
+
+```java
+@Test
+fun myTestCase() { ... }
+```
+
+## Espresso tests
+
+Making use of the interoperability of Java and Kotlin code we can blend Espresso into
+our Kotlin test classes in the blink of an eye. That's how we can benefit both from
+battle tested Espresso framework that makes testing UI components a bliss and
+all the gems that Kotlin provides as a language to make writing our _glue_ code and other
+testing logic as easy and clean as possible.
+
+As usual, we store our Espresso and other UI tests under `src/androidTest`, with a
+little discrepancy that instead of `java` subdirectory, we store our Kotlin source code in
+`kotlin` subdirectory. And in order to run Espresso tests we need to declare it as a dependency
```

Comment:
"with the minor difference that"
Since Kotlin is more relevant here, suggest switching the order: "with the minor difference that we store our Kotlin source code in the `kotlin` subdirectory instead of `java`"